.evergreen/config_generator/components/earthly.py

```diff
@@ -1,41 +1,171 @@
-            commands=[
-                subprocess_exec(
-                    "bash",
-                    args=[
-                        "tools/earthly.sh",
-                        f"+{target}",
-                        f"--env=${{{_ENV_PARAM_NAME}}}",
-                        f"--enable_sasl={sasl}",
-                    ],
-                    working_dir="mongoc",
-                    command_type=EvgCommandType.TEST,
-                )
-            ],
-            tags=[f"earthly", "pr-merge-gate"],
-            run_on=CONTAINER_RUN_DISTROS,
-        )
+_ENV_PARAM_NAME = "MONGOC_EARTHLY_ENV"
+"The name of the EVG expansion parameter used to key the Earthly build env"
 
+EnvKey = Literal["u22", "alpine3.18", "archlinux"]
+"Identifiers for environments. These correspond to special '*-env' targets in the Earthfile."
 
-#: A mapping from environment keys to the environment name.
-#: These correspond to special "*-env" targets in the Earthfile.
-ENVS = {
+_ENV_NAMES: dict[EnvKey, str] = {
     "u22": "Ubuntu 22.04",
     "alpine3.18": "Alpine 3.18",
     "archlinux": "Arch Linux",
 }
+"A mapping from environment keys to 'pretty' environment names"
+
+# Other options: SSPI (Windows only), AUTO (not reliably test-able without more environments)
+SASLOption = Literal["Cyrus", "off"]
+"Valid options for the SASL configuration parameter"
+TLSOption = Literal["LibreSSL", "OpenSSL", "off"]
+"Options for the TLS backend configuration parameter (AKA 'ENABLE_SSL')"
+CxxVersion = Literal["master", "r3.8.0"]
+"C++ driver refs that are under CI test"
+
+# A Unicode non-breaking space character
+_NBSP = "\N{No-break space}"
+
+
+class Configuration(NamedTuple):
+    """
+    Represent a complete set of argument values to give to the Earthly task
+    execution. Each field name matches the ARG in the Earthfile.
+
+    Adding/removing fields will add/remove dimensions on the task matrix.
+
+    The 'env' parameter is not encoded here, but is managed separately.
+
+    Keep this in sync with the 'PartialConfiguration' class defined below!
+    """
+
+    sasl: SASLOption
+    tls: TLSOption
+    test_mongocxx_ref: CxxVersion
+
+    @classmethod
+    def all(cls) -> Iterable[Configuration]:
+        """
+        Generate all configurations for all options of our parameters.
+        """
+        # Iter each configuration parameter:
+        fields: Iterable[tuple[str, type]] = get_type_hints(Configuration).items()
+        # Generate lists of pairs of parameter names their options:
+        all_pairs: Iterable[Iterable[tuple[str, str]]] = (
+            # Generate a (key, opt) pair for each option in parameter 'key'
+            [(key, opt) for opt in get_args(typ)]
+            # Over each parameter and type thereof:
+            for key, typ in fields
+        )
+        # Now generate the cross product of all alternative for all options:
+        matrix: Iterable[dict[str, Any]] = map(dict, itertools.product(*all_pairs))
```

Comment:
My concern isn't about the runtime test coverage, but that the rest of the project continues to build and install successfully when it is turned off (i.e. does code exist that assumes it is enabled?) At the moment, the tasks only execute the build without running the test suite, so they are very fast. It will be worthwhile to narrow down our task matrix if/when it becomes prohibitively expensive.